### PR TITLE
Add Rust and C++ API to borrow OpenGL textures

### LINKED ISSUE #977: Ability to provide custom texture (similar to QSGTextureProvider)
Hi!
I'm looking for a rust GUI library for my project. One of the critical requirements for me is a "custom render to texture" ability.  
Not a using slint as overlay for my opengl scene but an ability to insert my dynamically changed texture to the slint scene graph.
Similar class from Qt world: https://doc.qt.io/qt-5/qsgtextureprovider.html

Is it possible in slint? Thanks!

### PR DESCRIPTION
**Pull request by tronical**
This comes with a brand new example as well :-)

Screenshot

Closes #977

### CODE REVIEW

**Comment by tronical** [comment]
Sorry, no WASM version - not possible because glow doesn't expose the necessary APIs. We'd have to make glow a public dependency in Slint, which I'd like to avoid.

File: internal/core/graphics/image.rs [hunk 1]
```diff
@@ -780,4 +788,44 @@ pub(crate) mod ffi {
...
+///
+/// The ownership of the texture remains with the application. It is the application's responsibility to delete the texture
+/// when it is not used anymore.
+///
+/// Note that only 2D RGBA textures are supported.
+#[derive(Clone, Debug, PartialEq)]
+#[non_exhaustive]
+#[cfg(not(target_arch = "wasm32"))]
+#[repr(C)]
+pub struct BorrowedOpenGLTexture {
```
Comment (tronical) [inline internal/core/graphics/image.rs:810]: FWIW, I don't like this to be a top-level type, so my preference would be to move this into a `gpu` module (so `slint::gpu::BorrowedOpenGLTexture`.

File: internal/core/graphics/image.rs [hunk 2]
```diff
@@ -780,4 +788,44 @@ pub(crate) mod ffi {
...
     }
+
+    #[no_mangle]
+    pub unsafe extern "C" fn slint_image_compare_equal(image1: &Image, image2: &Image) -> bool {
+        return image1.eq(image2);
+    }
+}
+
+/// This structure contains fields to identify and render an OpenGL texture that Slint borrows from the applicatin code.
+/// Use this to embed a native OpenGL texture into a Slint scene, by creating the texture with [`Self::new()`] and
```
Comment (ogoffart) [inline internal/core/graphics/image.rs:799]: I wouldn't use Self here in the link as the user would use the fill class name

File: internal/core/graphics/image.rs [hunk 3]
```diff
@@ -780,4 +788,44 @@ pub(crate) mod ffi {
...
+/// converting it into an [`Image`] using `.into()`.
+///
+/// The ownership of the texture remains with the application. It is the application's responsibility to delete the texture
```
Comment (ogoffart) [inline internal/core/graphics/image.rs:802]: Wouldn't it be better to take ownership?
What happens if the application release the texture too early? 
And it seems easy to make a leak.

File: internal/core/graphics/image.rs (hunk 1 above)
Comment (ogoffart) [reply internal/core/graphics/image.rs:810]: Why do we even need this type? Couldn't we just have  Image::from_borrowed_opengl_texture(texture_id, size) ?

File: internal/renderers/skia/cached_image.rs [hunk 4]
```diff
@@ -77,6 +80,29 @@ pub(crate) fn as_skia_image(
...
+                _canvas.recording_context().as_mut().unwrap(),
+                &backend_texture,
+                skia_safe::gpu::SurfaceOrigin::TopLeft,
+                skia_safe::ColorType::RGBA8888,
+                skia_safe::AlphaType::Unpremul,
+                None,
+            )
+        },
+        #[cfg(not(skia_backend_opengl))]
+        ImageInner::BorrowedOpenGLTexture(..) => None,
```
Comment (ogoffart) [inline internal/renderers/skia/cached_image.rs:105]: In general, I wonder of we should have better error reporting.
Should we print something to the stderr?
Should we have a error callback? I think we should.
It doesn't have to be in this patch though.

File: internal/core/graphics/image.rs (hunk 3 above)
Comment (tronical) [reply internal/core/graphics/image.rs:802]: I don't think it's a good idea to offer transfer of ownership at the moment. I mean, for a texture that the application creates itself with glGenTextures that would be possible, but for example when retrieving textures from native APIs, they are only offered by borrow.

For example when mapping a gstreamer video frame to retrieve the GL texture handle, it's only borrowed. Or when retrieving the GL textures from CoreVideo on macOS via CVOpenGLTextureRef, the access to the handle is only borrowed.

In both cases the actual owner is reference counted, so we could try to map this via some type-erased arc that the user has to supply in as well to ensure that the reference count is dropped.

That said, neither Skia's Rust APIs support this at the moment (no mapping for a generic release procedure). Also note how Qt's public API for integrating external textures into QtQuick/RHI has abandoned the ownership concept in their API and leaves it with the application (it's worse there, the old code continues to compile, but the documentation just says that this isn't supported anymore).

So the only other option would be to ask the application to make a copy of the texture, which isn't just a call to some copy function, it requires building an FBO, a rendering pipeline, a vertex shader, a fragment shader, and then running the whole thing.

I'd prefer to make this API explicit about borrowing. If you want I can mark it as `unsafe`, but either way it's low-level. I'm against imposing the requirement to make a copy of the texture to the user.

To answer your question about what happens if the application releases the texture too early: Some weird graphical effects might happen (or often nothing), with OpenGL no crashes though.

We'll run into the same pattern of borrowing textures with Vulkan or Metal, and there a deletion that happens too early will result in crashes.

File: internal/core/graphics/image.rs (hunk 1 above)
Comment (tronical) [reply internal/core/graphics/image.rs:810]: Sure, but there will be a more in the future though - EGL external image, Vulkan, YUV. But if you don't mind then I'll add them all to `slint::Image`.

File: internal/renderers/skia/cached_image.rs (hunk 4 above)
Comment (tronical) [reply internal/renderers/skia/cached_image.rs:105]: I don't think this is the right place to handle errors or notify the application. If you create a `slint::Image` from an external OpenGL texture and use it with a renderer that doesn't support it, then right place to handle this is _before_ entering this code path. If we reach this code path, I think we should not render anything. But the right place to handle this is much earlier:

1. At compile or run-time, I want to make sure that a renderer is selected that supports this API.
2. At run-time I want to be able to check early on (in main()) what the capabilities of the graphics APIs in use are, and if the APIs the application needs are not supported, then do application specific error handling.

Trying to "handle" this at the time of rendering is futile IMO.


So two examples:

1. Suppose you're writing an application that shows maps and you know that the renderer requires OpenGL. I want to make sure that compile-time or very early at run-time that Slint supports the necessary features. If it's a run-time selection, it may be necessary to abort and inform the user. (Certainly not at rendering time).

2. Suppose you're writing a video player that should work everywhere. At start-up I need to check what APIs  are available: If OpenGL is available and I can efficiently create textures, then I may want to enable that code path in the renderer. Otherwise a fallback to creating slint::Image objects from pixel buffers may make more sense.


So in conclusion I think the correct way is to offer APIs to introspect graphics capabilities and optionally influence selection.

File: internal/core/graphics/image.rs (hunk 2 above)
Comment (tronical) [reply internal/core/graphics/image.rs:799]: Good idea.

File: internal/core/graphics/image.rs (hunk 1 above)
Comment (ogoffart) [reply internal/core/graphics/image.rs:810]: I think  several `Image::from_xyz`, `Image::from_hjk`, ... is actually more discoverable than many classes in a gpu modules. Don't you think?

File: internal/core/graphics/image.rs (hunk 3 above)
Comment (ogoffart) [reply internal/core/graphics/image.rs:802]: Ok.

> If you want I can mark it as unsafe, but either way it's low-level. I'm against imposing the requirement to make a copy of the texture to the user.

If the use of an invalid texture may result to undefined behavior, it definitively must be `unsafe`.  Otherwise if there is a failure mode (not showing the texture / warning in the console),  then it doesn't need to be unsafe

Comment (tronical) [reply internal/core/graphics/image.rs:802]: If you pass in garbage, then you'll get undefined behavior. It's not undefined in the main memory sense, but more in whatever the driver does. I don't mind starting with unsafe - at the very least this communicates that it's low-level API.

File: internal/core/graphics/image.rs (hunk 1 above)
Comment (tronical) [reply internal/core/graphics/image.rs:810]: I don't feel too strongly about it, so let's start with `from_xyz` - we can always revise this later.

**Review by ogoffart — APPROVED** [review]
